### PR TITLE
Polymer 3 Conversion

### PR DESCRIPTION
**Pull request by AdmiralCoco**

### CODE REVIEW

File: d2l-floating-buttons.js [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+
+@demo demo/floating-buttons.html d2l-floating-buttons
+*/
+import '../@polymer/polymer/polymer-legacy.js';
+
+import '../d2l-colors/d2l-colors.js';
+import '../@polymer/iron-media-query/iron-media-query.js';
+import './d2l-button-shared-styles.js';
+import { Polymer } from '../@polymer/polymer/lib/legacy/polymer-fn.js';
+import { html } from '../@polymer/polymer/lib/utils/html-tag.js';
```
Comment (dbatiste) [inline d2l-floating-buttons.js]: Interesting that it pulled in this tagged template function.   Adding `strip-whitespace` to the original would result in this component being handled similar to the rest.  Originally I was trying to be consistent with this (and the `strip-whitespace` is typically preferred anyway), but I'm not sure it makes any difference.

Comment (AdmiralCoco) [reply d2l-floating-buttons.js]: Oh I didn't notice that this component was missing `strip-whitespace`. I can add it and rerun the modulizer

**Comment by AdmiralCoco** [comment]
@awikkerink, sorry I'm not too familiar with Galen - what would be involved to update the tests to run with Polymer 3? 

On another note, I wasn't able to get the Galen tests to run locally. I get this error when I run `npm test` (before/after conversion):

```bash
[galen] > d2l-galen test test/acceptance/button.test.js -g factory:local -s
[galen]
[galen] ========================================
[galen] Test: phantomjs
[galen] ========================================
[galen] Nov 14, 2018 9:49:37 AM org.openqa.selenium.remote.DesiredCapabilities chrome
[galen] INFO: Using `new ChromeOptions()` is preferred to `DesiredCapabilities.chrome()`
[galen] org.mozilla.javascript.WrappedException: Wrapped java.lang.IllegalStateException: The path to the driver executable must be set by the webdriver.chrome.driver system property; for more information, see https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver. The latest version can be downloaded from http://chromedriver.storage.googleapis.com/index.html (<cmd>#37)
...
```

(I also tried to install the `chromedriver` dependency which got it to run, but it ended up hanging after a certain point)

**Comment by awikkerink** [comment]
@AdmiralCoco it would just involve deleting the `xEndpoint` variable and the `runTests('1.x', xEndpoint);` line.

Which platform are you running on?

**Comment by AdmiralCoco** [comment]
Ah okay, I see what you mean
I'm running this on Mac (Mojave)

**Comment by svanherk** [comment]
We're trying to make the script re-runnable in cases where that is reasonable.  I'm thinking we should apply the `strip-whitespace` changes, the linting changes, and the addition of `"prelint": "npm run build:sass",` to `master`/`hybrid` as well.  Then we can make and save a patch for the galen fixes and the additional linting needed for the "use single quotes" errors (these ones: https://github.com/BrightspaceUI/button/pull/141/commits/6d98e16c0986b27d97484b686e99cc01bc641361#diff-97f64c2e257692459b03a1c541bde15eR58)

**Comment by svanherk** [comment]
Similar to what we did for typography: https://github.com/Brightspace/polymer-3-converter/blob/master/patches/typography.demo.index.html.patch

File: package.json [hunk 2]
```diff
@@ -1,18 +1,21 @@
...
-    "postinstall": "polymer install --variants",
     "galen:local:run": "d2l-galen test test/acceptance/button.test.js -g factory:local -s",
     "galen:local:dump": "d2l-galen dump test/acceptance/button.test.js -d test/acceptance/dumps -g factory:local -s",
     "galen:sauce:run": "sc-run -- d2l-galen test test/acceptance/button.test.js -g factory:sauce -s -i SAUCE_USERNAME SAUCE_ACCESS_KEY TRAVIS_REPO_SLUG TRAVIS_BUILD_NUMBER -- -p4",
-    "lint": "npm run lint:wc && npm run lint:html",
-    "lint:html": "eslint *.html demo/*.html test/*.html test/acceptance/*.html test/acceptance/*.js",
-    "lint:wc": "polymer lint -i d2l-button.html d2l-floating-buttons.html d2l-button-subtle.html d2l-button-icon.html",
+    "prelint": "npm run build:sass",
+    "lint": "npm run lint:wc && npm run lint:js",
+    "lint:js": "eslint .--ignore-path reports/** --ext .js,.html test/**/*.js test/**/*.html demo/**/*.js demo/**/*.html",
```
Comment (svanherk) [inline package.json]: Was the `--ignore-path reports/**` needed even though we have `reports` in the `.eslintignore` file?

Comment (AdmiralCoco) [reply package.json]: Whoops, it seems like I left this from my attempts to fix linting

The issue that I ran into was when running `polymer lint`, it seems to lint the reports directory despite adding `filesToIgnore: ["reports/**/*"]` to `polymer.json` (as seen [here](https://github.com/Brightspace/brightspace-activity-details-ui/blob/030f393421cb982d85029a36be3051eae2cd76da/polymer.json#L15))

**Comment by AdmiralCoco** [comment]
> We're trying to make the script re-runnable in cases where that is reasonable. I'm thinking we should apply the `strip-whitespace` changes, the linting changes, and the addition of `"prelint": "npm run build:sass",` to `master`/`hybrid` as well. Then we can make and save a patch for the galen fixes and the additional linting needed for the "use single quotes" errors (these ones: [6d98e16#diff-97f64c2e257692459b03a1c541bde15eR58](https://github.com/BrightspaceUI/button/commit/6d98e16c0986b27d97484b686e99cc01bc641361#diff-97f64c2e257692459b03a1c541bde15eR58))

That makes sense - I'll open a new PR for these changes.

File: package.json (hunk 2 above)
Comment (svanherk) [reply package.json]: I've had lots of issues with running `polymer lint` locally, it always tries to lint things I don't want and then I get a memory error.  My solution has been to delete folders I don't want it to lint, but this isn't great :(

Comment (AdmiralCoco) [reply package.json]: Yeah I've also been doing that 😬 

I think the option to ignore paths was added, but from what I've seen there seems to have been some regression? https://github.com/Polymer/polymer-cli/issues/225

(I tried to downgrade `polymer-cli` but that didn't seem to work)

File: d2l-button-icon.js [hunk 3]
```diff
@@ -112,11 +112,26 @@ $_documentContainer.innerHTML = `<dom-module id="d2l-button-icon">
 				opacity: 0.5;
 			}
```
Comment (AdmiralCoco) [inline d2l-button-icon.js]: @svanherk, does this commit look OK for the patch file? Is there anything to add/remove?

Comment (svanherk) [reply d2l-button-icon.js]: Yeah I think this commit is good.  If we have to rerun the script the `FIXME` comments will be added back, but that's fine, we've just been leaving them in the other repos.

**Comment by svanherk** [comment]
Before merging I think we should run this in Sauce to make sure there aren't any issues.  Is there a way to manually do that, or should I checkout your fork and make a branch and push it to the repo?

**Comment by AdmiralCoco** [comment]
Sounds good. I can push the branch to the repo, I'll need to open a new PR for the Sauce tests to run right?

**Comment by awikkerink** [comment]
@AdmiralCoco correct. You can create a branch and the build will be attached to this PR automagically (github tracks builds by the hash)

**Comment by svanherk** [comment]
Still looks like it isn't running the Sauce tests, it thinks this is "Not a pull request and/or no secure environment variables, running headless tests..." 😕

**Comment by AdmiralCoco** [comment]
@svanherk I guess it thinks that it's still associated to my fork? I'll reopen this PR 

@awikkerink Yeah it looks like the failures are associated with `d2l-button-subtle`. It also appears that the same tests that fail with `shady-dom` pass with `shadow-dom` (except for one)

**Comment by awikkerink** [comment]
@AdmiralCoco correct. It appears that Galen is finding the span, but for some reason selenium thinks it's not visible. If the span is taken out of the slot in d2l-button-subtle, it works fine.

**Pull request closed without merging**